Build/libHttpClient.Android/src/main/java/com/xbox/httpclient/NetworkObserver.java

```diff
@@ -0,0 +1,145 @@
+            }
+        });
+    }
+
+    private static native void Log(String message);
+
+    static class NetworkDetails {
+        static String getNetworkDetails(Network network, ConnectivityManager cm) {
+            StringBuilder builder = new StringBuilder();
+
+            @Nullable NetworkCapabilities capabilities = cm.getNetworkCapabilities(network);
+            @Nullable LinkProperties linkProperties = cm.getLinkProperties(network);
+
+            return builder
+                .append("Network ")
+                .append(network.hashCode())
+                .append(":")
+                .append("\n  Capabilities: ")
+                .append(capabilities != null ? checkNetworkCapabilities(capabilities) : "Got null capabilities")
+                .append("\n  Link properties: ")
+                .append(linkProperties != null ? checkLinkProperties(linkProperties) : "Got null link properties")
+                .toString();
+        }
+
+        private static String checkNetworkCapabilities(@NotNull NetworkCapabilities capabilities) {
+            Map<String, Boolean> sections = new LinkedHashMap<>();
+
+            sections.put("isWifi", capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI));
+            sections.put("isBluetooth", capabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH));
+            sections.put("isCellular", capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR));
+            sections.put("isVpn", capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN));
+            sections.put("isEthernet", capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET));
+
+            sections.put("shouldHaveInternet", capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET));
+            sections.put("isNotVpn", capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN));
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                sections.put("internetWasValidated", capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED));
+            }
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                sections.put("isNotSuspended", capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_SUSPENDED));
+            }
+
+            return join(sections);
+        }
+
+        private static String checkLinkProperties(@NotNull LinkProperties properties) {
+            Map<String, Boolean> sections = new LinkedHashMap<>();
+
+            sections.put("hasProxy", properties.getHttpProxy() != null);
+
+            return join(sections);
+        }
+
+        private static String join(Map<String, Boolean> sections) {
+            StringBuilder builder = new StringBuilder();
+
+            boolean firstEntry = true;
+            for (Map.Entry<String, Boolean> entry : sections.entrySet()) {
+                if (firstEntry) {
```

Comment:
Can't you just test the string builder's length instead?

Reply:
Yeah, that would work too - that's a good thought!